src/ws/upgrade.rs

```diff
@@ -0,0 +1,270 @@
+use super::message::{Channel, Message};
+use crate::middleware::{BoxFuture, Middleware};
+use crate::next::Next;
+use crate::raise;
+use crate::request::RequestHead;
+use crate::response::Response;
+
+const DEFAULT_FRAME_SIZE: usize = 16 * 1024; // 16KB
+
+#[derive(Debug)]
+pub struct Request<State = ()> {
+    head: Arc<RequestHead<State>>,
+}
+
+pub struct Upgrade<F> {
+    config: Config,
+    limits: Limits,
+    listen: Arc<F>,
+}
+
+fn gen_accept_key(key: &[u8]) -> String {
+    let mut hasher = Hasher::new(&SHA1_FOR_LEGACY_USE_ONLY);
+
+    hasher.update(key);
+    hasher.update(b"258EAFA5-E914-47DA-95CA-C5AB0DC85B11");
+
+    base64_engine.encode(hasher.finish())
+}
+
+fn handle_error(error: &impl std::error::Error) {
+    if cfg!(debug_assertions) {
+        eprintln!("error(ws): {}", error);
+    }
+}
+
+async fn start<State, F, R>(listen: Arc<F>, mut head: Arc<RequestHead<State>>, builder: Builder)
+where
+    F: Fn(Channel, Request<State>) -> R + Send + Sync + 'static,
+    R: Future<Output = super::Result> + Send,
+{
+    let stream = {
+        let Some(head_mut) = Arc::get_mut(&mut head) else {
+            panic!("via::ws::upgrade: handshake already performed.");
+        };
+
+        let mut request = http::Request::new(());
+        swap(head_mut.extensions_mut(), request.extensions_mut());
+
+        let result = hyper::upgrade::on(&mut request).await;
+        swap(request.extensions_mut(), head_mut.extensions_mut());
+
+        match result {
+            Ok(upgraded) => builder.serve(TokioIo::new(upgraded)),
+            Err(error) => return handle_error(&error),
+        }
+    };
+
+    tokio::pin!(stream);
+
+    'session: loop {
+        let (sender, mut rx) = mpsc::channel(1);
+        let (tx, receiver) = mpsc::channel(1);
+        let mut listener = Box::pin(listen(
+            Channel::new(sender, receiver),
+            Request::new(Arc::clone(&head)),
+        ));
+
+        loop {
+            let flow = tokio::select! {
+                biased;
+
+                // The future returned from app code is ready.
+                result = listener.as_mut() => match result {
+                    Err(Continue(error)) => Continue(Some(error.into())),
+                    Err(Break(error)) => Break(Some(error.into())),
+                    Ok(_) => Break(None),
+                },
+
+                // Forward the outbound message to the stream.
+                Some(message) = coop::unconstrained(rx.recv()) => {
+                    let result = stream.feed(message.into()).await;
+
+                    coop::consume_budget().await;
+
+                    if let Err(error) = result {
+                        try_rescue_ws(error)
+                    } else {
+                        Continue(None)
+                    }
+                }
+
+                // Forward the incoming message to the channel.
+                Some(result) = stream.next() => {
+                    coop::consume_budget().await;
```

Comment:
This provides a bit of cushion by leaving a message on the stack if reading the next value out of the stream exceeds the budget for the task.

We may have to manually poll the listener future in order to ensure that the subsequent call to `tx.send` doesn't result in a deadlock but I'm pretty sure that yielding to the runtime is the right call here since the receiving half of the channel is unconstrained.

I may pull some people from the tokio group on discord in to the conversation here to confirm that this is the case. Otherwise, this will need a `poll_fn` to poll the listener on the subsequent line. In which case, I'd move the listener arm to the bottom of the select macro.